### ISSUE TITLE
Unable to create mock from trait with protected method

### ISSUE THREAD

**Issue by marcsaegesser** [paulbutcher/ScalaMock#114]
After upgrading from 3.0.1 to 3.2.2 I now get compiler errors when I'm trying to create a mock from a trait that contains protected methods. I'm using `org.scalamock.scalatest.MockFactory`. The error is

`object creation impossible, since method ... in trait ... of type ... is not defined`

Is there a way to make this work?

**Comment by nightscape** [comment]
I just ran into the same issue for private methods. Is there a known workaround?

**Comment by bkowalik** [comment]
**TL;DR;**
There is no solution for that. I assume that you want to mock `protected[pkg]`?

**Here is why:**
As you may know Java hasn't got such visibility levels like Scala has. Lets take simple trait:
```scala
package pkg;

trait Foo {
  protected[pkg] bar(a: Int): Int
  private[pkg] bar(a: Int): Int
}
```
For the sake of this example and comparison I have also added another method with `private[pkg]` visibility. Look what Scalac generates at the last phase of compilation (command `scalac -Xprint:23 Foo.scala`):
```scala
[[syntax trees at end of                     icode]] // Foo.scala
package pkg {
  abstract trait Foo extends Object {
    private[pkg] def bar(a: Int): Int;
    private[pkg] def baz(b: String): String
  }
}
```
Our protected method gets narrow modifier `private[pkg]`. If this is wired look at the JVM byte code (`javap -v -p -l pkg.Foo`):
```
{
  public abstract int bar(int);
    descriptor: (I)I
    flags: ACC_PUBLIC, ACC_ABSTRACT

  public abstract java.lang.String baz(java.lang.String);
    descriptor: (Ljava/lang/String;)Ljava/lang/String;
    flags: ACC_PUBLIC, ACC_ABSTRACT
}
```
As you can see both methods are public. WTF?! Traits as a pure interfaces are translated to Java interface so every modifier is erased because in bytecode below Java 8 you cannot have method implementations in interfaces. Even if you compile Scala with JDK8, Scala compiler is unable to emit such bytecode.

Here is how macro `mock[T]` expand for `Foo` type:
```scala
{
  final class $anon extends Foo {
    def <init>() = {
      super.<init>();
      ()
    };
    val mock$special$mockName = MockTest.this._factory.generateMockDefaultName("mock").name
  };
  new $anon()
}.asInstanceOf[Foo]
```

As you can see there is no method. WTF2?! That is because macros are applied on Scala AST where methods modifier is `private` and in ScalaMock you cannot mock private methods. No matter if you provide default implementation for any of this method (especially protected) you won't get access to it in ScalaMock. If you provide implementation, you would only force compiler to move your implementation to a class due to bytecode limitation that I've mentioned earlier.

I will look at it if there is anything in the AST that distinguish between private but really protected and real private because there are more information in the AST itself.

**Comment by barkhorn** [comment]
you can now mock these traits, but not define expectations on protected/private methods

**Comment by michaelahlers** [comment]
I've encountered a similar error on a _public_ member of a Java interface.

Error produced is:

```
[error] … object creation impossible, since method clone in trait Resource of type (x$1: String)org.apache.ivy.plugins.repository.Resource is not defined
[error]       val resource = mock[Resource]
```

And the [`clone` method is defined as](https://github.com/sbt/ivy/blob/2.4.0-rc1/src/java/org/apache/ivy/plugins/repository/Resource.java#L85-L92):

```java
public Resource clone(String cloneName);
```

**Comment by barkhorn** [comment]
Hi @michaelahlers - we tear off all methods that are defined in `java.lang.Object` as they cause fundamental problems or conflicts between assertions on mocks and the implementation of the library.
`clone` is one of those, even though it looks to be overloaded in this case with a `String` parameter.

https://github.com/paulbutcher/ScalaMock/blob/bfdc68010306980f7490c6366d46d79c1335ab6e/shared/src/main/scala/org/scalamock/clazz/MockMaker.scala#L308

**Comment by michaelahlers** [comment]
@barkhorn, thanks for the response! For anyone else who comes across this, an inoffensive workaround looks similar to:

```scala
abstract class MockableResource extends Resource {
  override def clone(cloneName: String): Resource = ???
}

// …

val resource: Resource = mock[MockableResource]
```

**Comment by barkhorn** [comment]
nice, thank you. this should probably be on the website :)
